### ISSUE TITLE
Failure to start node after upgrading to 1.2

### ISSUE THREAD

**Issue by Bouke** [citusdata/pg_auto_failover#261]
I've upgraded to version 1.2 across a cluster (monitor + 2 servers) and then restarted the monitor. Then I've taken down the secondary, and tried to bring it back up. This fails with the following error:

```
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  Monitor assigned new state "catchingup"
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  FSM transition from "demoted" to "catchingup": A new primary is available. First, try to rewind. If that fails, do a pg_basebackup.
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  The primary node returned by the monitor is acc-sql01.cb.local:5432
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  Rewinding PostgreSQL to follow new primary acc-sql01.cb.local:5432
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  pg_ctl: no server running
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  pg_ctl stop failed, but PostgreSQL is not running anyway
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  Running /usr/pgsql-12/bin/pg_rewind --target-pgdata "/var/lib/pgsql/12/data" --source-server "application_name=pgautofailover_standby_2 host=acc-sql01.cb.local port=5432 user=pgautofailover_replicator dbname=postgres" --progress ...
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  pg_rewind: fatal: could not connect to server: fe_sendauth: no password supplied
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 ERROR Failed to run pg_rewind: exit code 1
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 ERROR Failed to rewind old data directory
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 WARN  Failed to rewind demoted primary to standby, trying pg_basebackup instead
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  Initialising PostgreSQL as a hot standby
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  Target directory exists: "/var/lib/pgsql/12/data", stopping PostgreSQL
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  pg_ctl: no server running
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 INFO  pg_ctl stop failed, but PostgreSQL is not running anyway
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 ERROR Failed to ensure empty directory "": No such file or directory
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 ERROR Failed to become standby server, see above for details
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 ERROR Failed to transition from state "demoted" to state "catchingup", see above.
May 12 16:21:39 acc-sql02.cb.local pg_autoctl[1340]: 16:21:39 ERROR Failed to transition to state "catchingup", retrying...
```

Meanwhile on the primary while taking down the secondary:
```
May 12 16:19:05 acc-sql01.cb.local pg_autoctl[23597]: 16:19:05 INFO  Promoting postgres
May 12 16:19:05 acc-sql01.cb.local pg_autoctl[23597]: 16:19:05 INFO  Waiting for postgres to promote
May 12 16:19:06 acc-sql01.cb.local pg_autoctl[23597]: 16:19:06 INFO  Other node in the HA group is acc-sql02.cb.local:5432
May 12 16:19:06 acc-sql01.cb.local pg_autoctl[23597]: 16:19:06 WARN  Granting replication connection for "acc-sql02.cb.local" using authentication method "trust" although no replication password has been set
May 12 16:19:06 acc-sql01.cb.local pg_autoctl[23597]: 16:19:06 INFO  HINT: see `pg_autoctl config get replication.password`
May 12 16:19:06 acc-sql01.cb.local pg_autoctl[23597]: 16:19:06 INFO  Disabling synchronous replication
May 12 16:19:06 acc-sql01.cb.local pg_autoctl[23597]: 16:19:06 INFO  Transition complete: current state is now "stop_replication"
May 12 16:19:36 acc-sql01.cb.local pg_autoctl[23597]: 16:19:36 INFO  Monitor assigned new state "wait_primary"
May 12 16:19:36 acc-sql01.cb.local pg_autoctl[23597]: 16:19:36 INFO  FSM transition from "stop_replication" to "wait_primary": Confirmed promotion with the monitor
May 12 16:19:36 acc-sql01.cb.local pg_autoctl[23597]: 16:19:36 INFO  Setting default_transaction_read_only to off
May 12 16:19:36 acc-sql01.cb.local pg_autoctl[23597]: 16:19:36 INFO  Transition complete: current state is now "wait_primary"
```

It looks like 1.2 introduces a new auth model. We're using "trust" authentication mechanism. The logs seem to suggest I still need a password, something which I don't understand. Is there any upgrade documentation available?

**Comment by Bouke** [comment]
After adding `auth_method = none` to the config and `password =`, it tried to rewind, but failed:

```
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: connected to server
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: servers diverged at WAL location 4/A9007B50 on timeline 174
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: rewinding from last common checkpoint at 4/A9000148 on timeline 174
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: reading source file list
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: reading target file list
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: reading WAL in target
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: need to copy 56 MB (total source directory size is 250 MB)
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO      0/57513 kB (0%) copied
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  57513/57513 kB (100%) copied
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_rewind: creating backup label and updating control file
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  Writing recovery configuration to "/var/lib/pgsql/12/data/standby.signal"
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  Postgres is not running, starting postgres
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO   /usr/pgsql-12/bin/pg_ctl --pgdata /var/lib/pgsql/12/data --options "-p 5432" --options "-h *" --wait start
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 ERROR Failed to start PostgreSQL. pg_ctl start returned: 1
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 ERROR waiting for server to start....2020-05-12 15:18:22.462 GMT [7796] LOG:  syntax error in file "/var/lib/pgsql/12/data/postgresql-auto-failover-standby.conf" line 2, near token "'"
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 2020-05-12 15:18:22.464 GMT [7796] LOG:  syntax error in file "/var/lib/pgsql/12/data/postgresql-auto-failover-standby.conf" line 4, near token "["
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 2020-05-12 15:18:22.464 GMT [7796] LOG:  syntax error in file "/var/lib/pgsql/12/data/postgresql-auto-failover-standby.conf" line 12, near token "["
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 2020-05-12 15:18:22.464 GMT [7796] LOG:  syntax error in file "/var/lib/pgsql/12/data/postgresql-auto-failover-standby.conf" line 20, near token "["
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 2020-05-12 15:18:22.464 GMT [7796] LOG:  syntax error in file "/var/lib/pgsql/12/data/postgresql-auto-failover-standby.conf" line 26, near token "'"
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 2020-05-12 15:18:22.467 GMT [7796] FATAL:  configuration file "/var/lib/pgsql/12/data/postgresql.conf" contains errors
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]:  stopped waiting
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 ERROR Failed to start postgres after rewind
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 WARN  Failed to rewind demoted primary to standby, trying pg_basebackup instead
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  Initialising PostgreSQL as a hot standby
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  Target directory exists: "/var/lib/pgsql/12/data", stopping PostgreSQL
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_ctl: no server running
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 INFO  pg_ctl stop failed, but PostgreSQL is not running anyway
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 ERROR Failed to ensure empty directory "": No such file or directory
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 ERROR Failed to become standby server, see above for details
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 ERROR Failed to transition from state "demoted" to state "catchingup", see above.
May 12 17:18:22 acc-sql02.cb.local pg_autoctl[7517]: 17:18:22 ERROR Failed to transition to state "catchingup", retrying...
```

Further attempts also failed:
```
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  Monitor assigned new state "catchingup"
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  FSM transition from "demoted" to "catchingup": A new primary is available. First, try to rewind. If that fails, do a pg_basebackup.
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  The primary node returned by the monitor is acc-sql01.cb.local:5432
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  Rewinding PostgreSQL to follow new primary acc-sql01.cb.local:5432
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  pg_ctl: no server running
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  pg_ctl stop failed, but PostgreSQL is not running anyway
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  Running /usr/pgsql-12/bin/pg_rewind --target-pgdata "/var/lib/pgsql/12/data" --source-server "application_name=pgautofailover_standby_2 host=acc-sql01.cb.local port=5432 user=pgautofailover_replicator dbname=postgres" --progress ...
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  pg_rewind: connected to server
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  pg_rewind: fatal: target server must be shut down cleanly
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 ERROR Failed to run pg_rewind: exit code 1
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 ERROR Failed to rewind old data directory
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 WARN  Failed to rewind demoted primary to standby, trying pg_basebackup instead
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  Initialising PostgreSQL as a hot standby
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  Target directory exists: "/var/lib/pgsql/12/data", stopping PostgreSQL
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  pg_ctl: no server running
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 INFO  pg_ctl stop failed, but PostgreSQL is not running anyway
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 ERROR Failed to ensure empty directory "": No such file or directory
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 ERROR Failed to become standby server, see above for details
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 ERROR Failed to transition from state "demoted" to state "catchingup", see above.
May 12 17:18:27 acc-sql02.cb.local pg_autoctl[7517]: 17:18:27 ERROR Failed to transition to state "catchingup", retrying...
```

The error message seemed to suggest to remove the data directory, but that made matters worse:
```
May 12 17:20:39 acc-sql02.cb.local pg_autoctl[7517]: 17:20:39 ERROR pg_ctl: directory "/var/lib/pgsql/12/data" is not a database cluster directory
```

So next, I'll try to remove this node from the monitor. I'm on `acc-sql02`:
```
$ pg_autoctl drop node
17:29:45 INFO  Found a stale pidfile at "/tmp/pg_autoctl/var/lib/pgsql/12/data/pg_autoctl.pid"
17:29:45 WARN  Removing the stale pid file "/tmp/pg_autoctl/var/lib/pgsql/12/data/pg_autoctl.pid"
17:29:45 WARN  Skipping node removal from the monitor: state file "/var/lib/pgsql/.local/share/pg_autoctl/var/lib/pgsql/12/data/pg_autoctl.state" does not exist
17:29:45 INFO  Stopping PostgreSQL at "/var/lib/pgsql/12/data"
17:29:45 ERROR could not change directory to "/home/bhaarsma": Permission denied
17:29:45 ERROR pg_ctl: directory "/var/lib/pgsql/12/data" is not a database cluster directory
17:29:45 INFO  Stopping PostgreSQL server failed. pg_ctl status returned: 4
17:29:45 ERROR could not change directory to "/home/bhaarsma": Permission denied
17:29:45 ERROR pg_ctl: directory "/var/lib/pgsql/12/data" is not a database cluster directory
17:29:45 ERROR Failed to stop PostgreSQL at "/var/lib/pgsql/12/data"

$ pg_autoctl show state --formation acceptance
              Name |   Port | Group |  Node |     Current State |    Assigned State
-------------------+--------+-------+-------+-------------------+------------------
acc-sql01.cb.local |   5432 |     0 |     1 |      wait_primary |      wait_primary
acc-sql02.cb.local |   5432 |     0 |     2 |           demoted |        catchingup
```

**Comment by DimCitus** [comment]
Hi @Bouke ; thanks for opening this issue. Several things are hard for me to understand here. The `auth` parameter impacts mainly the creation of the HBA entries, which are already in place when you upgrade.

The next error message is `ERROR Failed to ensure empty directory "": No such file or directory` which indicates you don't have a `replication.backup_directory` set on this instance, and that should be automatic and should have been set in previous versions already. Did you maybe skip a version and upgrade from 1.0.6 straight to 1.2, without using 1.1?

About versions, we just release 1.3 and fixed several usability bugs found in 1.2, so I would advice to use 1.3 already. It can be seen as a better 1.2.

Then we have `LOG:  syntax error in file "/var/lib/pgsql/12/data/postgresql-auto-failover-standby.conf" line 4, near token "["` from the Postgres logs themselves and that's very confusing, can you paste the content of this file?

**Comment by Bouke** [comment]
> Several things are hard for me to understand here.

The upgrade was very confusing to me, and my main concern was getting it up and running again.  I tried my best to recover all the steps I've taken. Some things I might've done could be ill-adviced or even counterproductive.

>  The auth parameter impacts mainly the creation of the HBA entries, which are already in place when you upgrade.

I got into the situation where the secondary couldn't replicate from the primary due to an authentication issue. Eventually I think the problem was that on the monitor the `username` was set to `autoctl_node` where changing it to `postgres` worked. The monitor is running under the `postgres` user, and that user is allowed to access any database on the system. This used to work with the `autoctl_node` user, but somehow it didn't anymore:

```
local   pg_auto_failover        autoctl_node            trust
```
> which indicates you don't have a replication.backup_directory set on this instance, and that should be automatic and should have been set in previous versions already.

It was indeed empty on our install. We started with 1.0.5 I think. Since I've setup the node as new, the value has been set correctly.

> Did you maybe skip a version and upgrade from 1.0.6 straight to 1.2, without using 1.1?

We upgraded from 1.0.6 to 1.2 indeed. There's no 1.1 in the repo, could it be missing? The changelog doesn't mention 1.1, which is somewhat confusing.
```
yum --showduplicates list pg_auto_failover_12 | expand
Last metadata expiration check: 0:01:48 ago on Wed 13 May 2020 03:57:42 PM CEST.
Installed Packages
pg_auto_failover_12.x86_64           1.2-1.rhel8             @yum.postgresql.org
Available Packages
pg_auto_failover_12.x86_64           1.0.4-1.rhel8           yum.postgresql.org
pg_auto_failover_12.x86_64           1.0.5-1.rhel8           yum.postgresql.org
pg_auto_failover_12.x86_64           1.2-1.rhel8             yum.postgresql.org
```

As 1.3 is not (yet) available in the repo, we'll be using 1.2 for now.

> Then we have LOG: syntax error in file "/var/lib/pgsql/12/data/postgresql-auto-failover-standby.conf" line 4, near token "[" from the Postgres logs themselves and that's very confusing, can you paste the content of this file?

I have since removed that file and failed to create a backup.

**Comment by DimCitus** [comment]
Closing for triaging purposes, feel free to re-open if the issue still applies to current version of pg_auto_failover.
